crates/ra_hir/src/expr/validation.rs

```diff
@@ -0,0 +1,67 @@
+use rustc_hash::FxHashSet;
+use crate::{
+    expr::AstPtr,
+    HirDatabase,
+    Function,
+    Name,
+    diagnostics::{DiagnosticSink, MissingFields},
+    adt::AdtDef,
+    Path
+};
+
+use ra_syntax::ast::{AstNode, NamedFieldList};
+
+use super::{Expr, StructLitField, ExprId};
+
+pub(crate) fn validate_body(func: &Function, db: &impl HirDatabase, sink: &mut DiagnosticSink) {
+    let body = func.body(db);
+    for e in body.exprs() {
+        match e {
+            (id, Expr::StructLit { path, fields, spread }) => {
+                validate_struct_literal(id, path, fields, spread, func, db, sink)
+            }
+            _ => (),
+        }
+    }
+}
+
+fn validate_struct_literal(
+    id: ExprId,
+    _path: &Option<Path>,
+    fields: &Vec<StructLitField>,
+    spread: &Option<ExprId>,
+    func: &Function,
+    db: &impl HirDatabase,
+    sink: &mut DiagnosticSink,
+) {
+    if let Some(_) = spread {
+        return;
+    }
+    let file_id = func.source(db).0;
+    let source_map = func.body_source_map(db);
```

Comment:
I think it's better to avoid fetching sources unless we actually have an error. That should allow to better cache absence of diagnostics, if we decide to do that